server/src-lib/Hasura/GraphQL/Transport/WebSocket/Server.hs

```diff
@@ -145,13 +161,28 @@ createServerApp
   -- user provided handlers
   -> WSHandlers a
   -- aka WS.ServerApp
-  -> WS.PendingConnection -> IO ()
-createServerApp (WSServer logger@(L.Logger writeLog) connMap) wsHandlers pendingConn = do
+  -> WS.PendingConnection
+  -> IO ()
+createServerApp (WSServer logger@(L.Logger writeLog) connMap shutdownVar) wsHandlers
+                pendingConn = do
+
   wsId <- WSId <$> UUID.nextRandom
   writeLog $ WSLog wsId EConnectionRequest
-  let reqHead = WS.pendingRequest pendingConn
-  onConnRes <- _hOnConn wsHandlers wsId reqHead
-  either (onReject wsId) (onAccept wsId) onConnRes
+  serverStatus <- STM.readTVarIO shutdownVar
+  case serverStatus of
+    AcceptingConns -> do
+      let reqHead = WS.pendingRequest pendingConn
+      onConnRes <- _hOnConn wsHandlers wsId reqHead
+      either (onReject wsId) (onAccept wsId) onConnRes
```

Comment:
Is there a lingering race here? If `_wssShutdown` is set to `ShuttingDown` after `shutdownVar` is read but before `onAccept` adds the connection to the map, then I think it still won’t be closed. I think, fundamentally, we need to arrange for the insertion itself to be guarded by a `serverStatus` check.

This is tricky to get right (I didn’t think about it when I was reviewing this yesterday, either, so we’ve both gotten it wrong!), so maybe we should change `WSServer` to the following:

```haskell
data WSServer a
  = WSServer
  { _wssLogger :: !L.Logger
  , _wssStatus :: !(TVar (ServerStatus a))
  }

data ServerStatus a
  = AcceptingConns !(ConnMap a)
  | ShuttingDown
```

I think that should make a race of this sort essentially unrepresentable, assuming `shutdown` does the obvious thing of setting the state to `ShuttingDown` and reading the current connections in a single, atomic operation.

Reply:
Very interesting, and very tricky indeed. I will play with that idea for a bit since it hasn't yet clicked for me, but at least it is obvious now that there are more potential races in the current solution. Thanks for noticing!